consensus/leader.go

```diff
@@ -173,73 +181,105 @@ func (consensus *Consensus) onPrepare(msg *msg_pb.Message) {
 
 func (consensus *Consensus) onCommit(msg *msg_pb.Message) {
 	recvMsg, err := ParseFBFTMessage(msg)
-
+	log := consensus.getLogger()
 	if err != nil {
 		consensus.getLogger().Debug().Err(err).Msg("[OnCommit] Parse pbft message failed")
 		return
 	}
 
-	// let it handle its own logs
-	// TODO(Edgar)(TEMP DISABLE while testing double sign)
-	if !consensus.onCommitSanityChecks(recvMsg) {
+	// NOTE let it handle its own log
+	if !consensus.isRightBlockNumAndViewID(recvMsg) {
 		return
 	}
 
 	consensus.mutex.Lock()
 	defer consensus.mutex.Unlock()
 
+	if key := (bls.PublicKey{}); consensus.couldThisBeADoubleSigner(recvMsg) {
+		if alreadyCastBallot := consensus.Decider.ReadBallot(
+			quorum.Commit, recvMsg.SenderPubkey,
+		); alreadyCastBallot != nil {
+			for _, blk := range consensus.FBFTLog.GetBlocksByNumber(recvMsg.BlockNum) {
+				alreadyCastBallot.SignerPubKey.ToLibBLSPublicKey(&key)
+				if recvMsg.SenderPubkey.IsEqual(&key) {
+					signed := blk.Header()
+					areHeightsEqual := signed.Number().Uint64() == recvMsg.BlockNum
+					areViewIDsEqual := signed.ViewID().Uint64() == recvMsg.ViewID
+					areHeadersEqual := bytes.Compare(
+						signed.Hash().Bytes(), recvMsg.BlockHash.Bytes(),
+					) == 0
+					// If signer already signed, and the block height is the same
+					// and the viewID is the same, then we need to verify the block
+					// hash, and if block hash is different, then that is a clear
+					// case of double signing
+					if areHeightsEqual && areViewIDsEqual && !areHeadersEqual {
+						var doubleSign bls.Sign
+						if err := doubleSign.Deserialize(recvMsg.Payload); err != nil {
+							log.Err(err).Str("msg", recvMsg.String()).
+								Msg("could not deserialize potential double signer")
+							return
+						}
+
+						curHeader := consensus.ChainReader.CurrentHeader()
+						committee, err := consensus.ChainReader.ReadShardState(curHeader.Epoch())
+						if err != nil {
+							log.Err(err).
+								Uint32("shard", consensus.ShardID).
+								Uint64("epoch", curHeader.Epoch().Uint64()).
+								Msg("could not read shard state")
+							return
+						}
+						offender := *shard.FromLibBLSPublicKeyUnsafe(recvMsg.SenderPubkey)
+						addr, err := committee.FindCommitteeByID(
+							consensus.ShardID,
+						).AddressForBLSKey(offender)
+
+						if err != nil {
+							log.Err(err).Str("msg", recvMsg.String()).
+								Msg("could not find address for bls key")
+							return
+						}
+
+						now := big.NewInt(time.Now().UnixNano())
+
+						go func(reporter common.Address) {
+							evid := slash.Evidence{
+								ConflictingBallots: slash.ConflictingBallots{
+									*alreadyCastBallot,
+									votepower.Ballot{
+										offender,
+										recvMsg.BlockHash,
+										common.Hex2Bytes(doubleSign.SerializeToHexStr()),
+									}},
+								Moment: slash.Moment{
+									Epoch:        curHeader.Epoch(),
```

Comment:
make sure the epoch/height/viewid/shardID are the exact ones of the double sign. Recording the current ones is not useful. So maybe use those from "signed" and "recvMsg"